File: Java-Core/src/main/java/prompto/value/TimeRange.java
package prompto.value;

import prompto.intrinsic.PromptoRange;
import prompto.intrinsic.PromptoTime;
import prompto.type.TimeType;



public class TimeRange extends RangeBase<Time> {

	static class PromptoTimeRange extends PromptoRange<Time> {

		public PromptoTimeRange(prompto.value.Time low, prompto.value.Time high) {
			super(low, high);
		}

		@Override
		public prompto.value.Time getItem(long item) {
			PromptoTime result = low.getValue().plusSeconds(item-1);
			if(result.isAfter(high.getValue()))
				throw new IndexOutOfBoundsException();
			return new prompto.value.Time(result);
		}
		
		@Override
		public PromptoTimeRange slice(long first, long last) {
			last = adjustLastSliceIndex(last);
			return new PromptoTimeRange(getItem(first), getItem(last));
		}

		@Override
		public long length() {
			return 1 + (high.getMillisOfDay() - low.getMillisOfDay())/1000;
		}
		

	}
	
	public TimeRange(Time left, Time right) {
		this(new PromptoTimeRange(left, right));
	}
	
	
	public TimeRange(PromptoRange<Time> range) {
		super(TimeType.instance(), range);
	}


	@Override
	public int compare(Time o1, Time o2) {
		return o1.compareTo(o2);
	}

	@Override
	public RangeBase<Time> newInstance(PromptoRange<Time> range) {
		return new TimeRange(range);
	}


}

File: Java-CodeServer/src/main/resources/js/prompto/value/TimeRange.js
var Range = require("./Range").Range;
var Time = require("./Time").Time;
var TimeType = null;

exports.resolve = function() {
    TimeType = require("../type/TimeType").TimeType;
};

function TimeRange(left, right) {
	Range.call(this, TimeType.instance, left, right);
	return this;
}

TimeRange.prototype = Object.create(Range.prototype);
TimeRange.prototype.constructor = TimeRange;


TimeRange.prototype.size = function() {
	return 1 + (this.high.value.valueOf() - this.low.value.valueOf())/1000;
};

TimeRange.prototype.compare = function(o1, o2) {
	return o1.cmp(o2);
};

TimeRange.prototype.getItem = function(index) {
	var result = this.low.value.valueOf() + (index-1)*1000;
	if(result>this.high.value.valueOf()) {
		throw new IndexOutOfBoundsException();
	}
	return new Time(new Date(result));
};

/*
@Override
public Range<Time> newInstance(Time left, Time right) {
	return new TimeRange(left, right);
}
*/

exports.TimeRange = TimeRange;
